File: dart_core/lib/src/exceptions.dart
/// Base exception class for all Telebirr SDK errors.
///
/// All SDK-specific exceptions extend this class, allowing callers to
/// catch any SDK error with a single `catch` clause on [TelebirrException].
///
/// Example:
/// ```dart
/// try {
///   await payments.charge(params);
/// } on TelebirrException catch (e) {
///   print('Error ${e.code}: ${e.message}');
/// }
/// ```
class TelebirrException implements Exception {
  /// A human-readable error message describing what went wrong.
  final String message;

  /// A machine-readable error code identifying the error category.
  ///
  /// Defaults to `'UNKNOWN'` when no specific code is provided.
  final String code;

  /// Creates a [TelebirrException] with the given [message] and optional [code].
  TelebirrException(this.message, {this.code = 'UNKNOWN'});

  @override
  String toString() => 'TelebirrException($code): $message';
}

/// Exception thrown when input validation fails.
///
/// This occurs when required parameters are missing, have invalid formats,
/// or violate business rules (e.g., a non-positive refund amount).
///
/// Example:
/// ```dart
/// try {
///   await payments.refund(params);
/// } on ValidationException catch (e) {
///   print('Validation error: ${e.message}');
/// }
/// ```
class ValidationException extends TelebirrException {
  /// Creates a [ValidationException] with the given [message] and optional [code].
  ValidationException(super.message, {super.code = 'VALIDATION'});
}

/// Exception thrown when request signing or signature verification fails.
///
/// This can happen due to an invalid private key, malformed PEM, or
/// errors during the RSA-PSS signing process.
///
/// Example:
/// ```dart
/// try {
///   await payments.charge(params);
/// } on SigningException catch (e) {
///   print('Signing error: ${e.message}');
/// }
/// ```
class SigningException extends TelebirrException {
  /// Creates a [SigningException] with the given [message] and optional [code].
  SigningException(super.message, {super.code = 'SIGNING'});
}

/// Exception thrown when an HTTP request to the Telebirr API fails.
///
/// This covers connection errors, non-2xx HTTP status codes, and
/// malformed response bodies.
///
/// Example:
/// ```dart
/// try {
///   await payments.charge(params);
/// } on NetworkException catch (e) {
///   print('Network error: ${e.message}');
/// }
/// ```
class NetworkException extends TelebirrException {
  /// Creates a [NetworkException] with the given [message] and optional [code].
  NetworkException(super.message, {super.code = 'NETWORK'});
}

/// Exception thrown when the SDK is configured with mismatched environment keys.
///
/// For example, using a test key (`sk_test_*`) with the production environment
/// or a live key (`sk_live_*`) with the sandbox environment.
///
/// Example:
/// ```dart
/// try {
///   final sdk = Telebirr(config);
/// } on EnvironmentException catch (e) {
///   print('Config error: ${e.message}');
/// }
/// ```
class EnvironmentException extends TelebirrException {
  /// Creates an [EnvironmentException] with the given [message] and optional [code].
  EnvironmentException(super.message, {super.code = 'ENVIRONMENT'});
}

File: ts-core/src/exceptions.ts
/**
 * Base error class for all Telebirr SDK errors.
 *
 * All SDK-specific exceptions extend this class, making it easy to
 * catch any SDK error with a single `instanceof` check.
 *
 * @example
 * ```ts
 * try {
 *   await telebirr.payments.charge(params);
 * } catch (err) {
 *   if (err instanceof TelebirrError) {
 *     console.error(err.code, err.message);
 *   }
 * }
 * ```
 */
export class TelebirrError extends Error {
  /** Machine-readable error code (e.g. `'VALIDATION_ERROR'`, `'NETWORK_ERROR'`). */
  code: string;

  /**
   * @param message - Human-readable error description.
   * @param code - Machine-readable error code. Defaults to `'UNKNOWN'`.
   */
  constructor(message: string, code: string = 'UNKNOWN') {
    super(message);
    this.name = 'TelebirrError';
    this.code = code;
  }
}

/**
 * Thrown when input validation fails (e.g. missing required fields, invalid amounts).
 *
 * @example
 * ```ts
 * // Throws ValidationError: "Amount must be a positive number"
 * await telebirr.payments.charge({ amount: '-5', title: 'Test' });
 * ```
 */
export class ValidationError extends TelebirrError {
  /** @param message - Description of the validation failure. */
  constructor(message: string) {
    super(message, 'VALIDATION_ERROR');
    this.name = 'ValidationError';
  }
}

/**
 * Thrown when request signing or signature verification fails.
 *
 * Typically caused by an invalid or mismatched private key.
 */
export class SigningError extends TelebirrError {
  /** @param message - Description of the signing failure. */
  constructor(message: string) {
    super(message, 'SIGNING_ERROR');
    this.name = 'SigningError';
  }
}

/**
 * Thrown when an HTTP request to the Telebirr API fails,
 * including network-level errors and non-2xx responses.
 *
 * @example
 * ```ts
 * try {
 *   await telebirr.payments.query({ merchOrderId: '123' });
 * } catch (err) {
 *   if (err instanceof NetworkError) {
 *     console.error('HTTP failure:', err.message);
 *   }
 * }
 * ```
 */
export class NetworkError extends TelebirrError {
  /** @param message - Description of the network failure. */
  constructor(message: string) {
    super(message, 'NETWORK_ERROR');
    this.name = 'NetworkError';
  }
}

/**
 * Thrown when the SDK detects a mismatch between the key and environment,
 * such as using a test key with the production endpoint or vice versa.
 *
 * @example
 * ```ts
 * // Throws EnvironmentError: "Test key cannot be used with PRODUCTION environment"
 * new Telebirr({ ...config, environment: 'PRODUCTION', fabricAppId: 'sk_test_xxx' });
 * ```
 */
export class EnvironmentError extends TelebirrError {
  /** @param message - Description of the environment mismatch. */
  constructor(message: string) {
    super(message, 'ENVIRONMENT_ERROR');
    this.name = 'EnvironmentError';
  }
}
